File: database/src/main/java/dreammaker/android/expensetracker/database/IExpenseDatabase.java
package dreammaker.android.expensetracker.database;

import java.util.concurrent.Callable;

import dreammaker.android.expensetracker.database.dao.AccountDao;
import dreammaker.android.expensetracker.database.dao.GroupDao;
import dreammaker.android.expensetracker.database.dao.HistoryDao;

public interface IExpenseDatabase {

    void runInTransaction(Runnable task);

    <V> V runInTransaction(Callable<V> task);

    HistoryDao getHistoryDao();

    AccountDao getAccountDao();

    GroupDao getGroupDao();
}

File: database/src/main/java/dreammaker/android/expensetracker/database/IExpenseDatabase.kt
package dreammaker.android.expensetracker.database

import dreammaker.android.expensetracker.database.dao.AccountDao
import dreammaker.android.expensetracker.database.dao.AnalyticsDao
import dreammaker.android.expensetracker.database.dao.GroupDao
import dreammaker.android.expensetracker.database.dao.HistoryDao

interface IExpenseDatabase {

    fun <V> runInTransaction(task: suspend ()->V): V

    val historyDao: HistoryDao

    val accountDao: AccountDao

    val groupDao: GroupDao

    val analyticsDao: AnalyticsDao
}
